### PR TITLE
fix: implement solution for felt representation mismatch between rust and miden

### PR DESCRIPTION
**Pull request by bitwalker**
This PR implements fixes for the following items:

- #169
- #239 
- #228 
- Most of the issues in #74 (a few of the 64-bit ops may still be unimplemented, but most are now)

See the commit messages for specific changes, but in general:

- Felt is now represented as f32 in Rust
- Felt is now 32-bits in the HIR type system
- Conversions to/from Felt have been rewritten accordingly
- Lowerings for 64-bit ops have been reviewed/tweaked, and in some cases replaced with stdlib calls. There were bugs found during this process that may be related to other ongoing issues being diagnosed
- Lowerings for u128 have been implemented (only for a limited subset of ops currently)
- The codegen crate now links in the Miden stdlib when using `MasmCompiler`, other areas of the codebase that do not test via that interface can make use of the same primitives it uses, but I have not done that here.

@greenhat One issue that will come up once this is merged (temporarily) is that the `cargo-miden` tests will fail until we update the `rust-templates` repo to point to this commit. The templates refer to a version of the compiler that uses the 64-bit felt representation, which will cause an error. I tested locally by manually specifying my local `rust-templates` clone, where I used a path dependency to refer to the version of the compiler I was working on, so this PR has a working test suite - it just won't work in CI until we make the changes I just described. Unfortunately I don't know of any way around it in this case.

We should probably merge this before our other PRs - I suspect some of the issues I found and fixed will help in your PR stack with testing and what not, and the spills PR needs these changes as my tests depend on the 32-bit felt layout.

### CODE REVIEW

**Comment by greenhat** [comment]
> We should probably merge this before our other PRs - I suspect some of the issues I found and fixed will help in your PR stack with testing and what not, and the spills PR needs these changes as my tests depend on the 32-bit felt layout.

Sounds good! I'm still in the middle of the spills PR review. I could switch to this one and then resume spills PR review.

**Comment by bitwalker** [comment]
@greenhat I'd prioritize this one first, but if you are in the middle of a tricky part of the spills PR don't feel like you have to switch in the middle of that - just whenever you find a convenient time to context switch.

**Comment by greenhat** [comment]
> @greenhat One issue that will come up once this is merged (temporarily) is that the `cargo-miden` tests will fail until we update the `rust-templates` repo to point to this commit. The templates refer to a version of the compiler that uses the 64-bit felt representation, which will cause an error. I tested locally by manually specifying my local `rust-templates` clone, where I used a path dependency to refer to the version of the compiler I was working on, so this PR has a working test suite - it just won't work in CI until we make the changes I just described. Unfortunately I don't know of any way around it in this case.

What I did in such cases:
1. Make a PR in `rust-templates` repo and use a last commit from this PR as a dependency;
2. Use a commit from step 1 PR in cargo extension in this PR for fetching `rust-templates`;
3. Update the commit hashes if merging these PRs renders some commits unavailable.

It's a bit of extra work, but it keeps CI green.

File: frontend-wasm/src/code_translator/mod.rs [hunk 1]
```diff
@@ -327,10 +327,6 @@ pub fn translate_operator(
             // https://www.w3.org/TR/wasm-core-1/#op-isub
             state.push1(builder.ins().sub_wrapping(arg1, arg2, span));
         }
-        Operator::F64Sub => {
```
Comment (greenhat) [inline frontend-wasm/src/code_translator/mod.rs:330]: Good catch! Thanks!

**Comment by bitwalker** [comment]
@greenhat I'll merge this PR once CI is green, but if you are curious, take a look at the latest commit to see how I set up usage of the new rust-templates options

**Pull request merged**